File: leetcode/java/LEET687.java
/**
 * LEET687
 */
public class LEET687 {    
//   Definition for a binary tree node.
    public static class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode() {}
        TreeNode(int val) { this.val = val; }
        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }
    public static void main(String[] args){
        Integer[] root = {5,4,5,1,1,5};
        TreeNode[] nodes = new TreeNode[root.length];
        for(int i=0;i<root.length;i++){
            if(root[i] == null){
                nodes[i] = null;
            } else {
                nodes[i] = new TreeNode(root[i]);
            }
        }

        for(int i=0;i<nodes.length;i++){
            if(nodes[i] == null) continue;
            TreeNode cur_node = nodes[i];
            int left = i*2+1, right = i*2+2;

            if(left < nodes.length && nodes[left] != null){
                cur_node.left = nodes[left];
            }
            if(right < nodes.length && nodes[right] != null){
                cur_node.right = nodes[right];
            }
        }
        System.out.println(longestUnivaluePath(nodes[0]));
    }
    static int longest = 0;
    public static int longestUnivaluePath(TreeNode root){
        dfs(root);
        return longest;
    }
    public static int dfs(TreeNode node){
        if(node == null){
            return 0;
        }
        int left = dfs(node.left);
        int right = dfs(node.right);

        if(node.left != null && node.left.val == node.val)
            left++;
        else
            left = 0;
        
        if(node.right != null && node.right.val == node.val)
            right++;
        else
            right = 0;

        longest = Math.max(longest, left+right);
        return Math.max(left, right);
    }
}

File: 0_3.leetcode/python/LEET687.py
from typing import *

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

longest = 0
def longestUnivaluePath(root: TreeNode) -> int:
    def dfs(node):
        global longest
        if not node:
            return 0 
        
        left = dfs(node.left)
        right = dfs(node.right)

        if node.left and node.left.val == node.val:
            left += 1
        else:
            left = 0
        
        if node.right and node.right.val == node.val:
            right += 1
        else:
            right = 0
        
        longest = max(longest, left+right)
        return max(left, right)
    dfs(root)
    return longest
def solve():
    root = [1,4,5,4,4,5]
    nodes = [TreeNode(r) if r is not None else None for r in root]
    
    for i in range(len(nodes)):
        node = nodes[i]
        left, right = i * 2 + 1, i * 2 + 2 

        if left < len(nodes):
            if nodes[left] is None: continue
            node.left = nodes[left]
        if right < len(nodes):
            if nodes[right] is None: continue
            node.right = nodes[right]

    res = longestUnivaluePath(nodes[0])
    print(res)

solve()